lua/nvim-treesitter/textobjects/move.lua

```diff
@@ -70,16 +74,157 @@ local function move(query_strings, forward, start, winid)
 end
 
 M.goto_next_start = function(query_strings)
-  move(query_strings, "forward", "start")
+  local args = { query_strings, "forward", "start" }
+  move(unpack(args))
+  M.last_move = { func = move, args = args }
 end
 M.goto_next_end = function(query_strings)
-  move(query_strings, "forward", not "start")
+  local args = { query_strings, "forward", not "start" }
+  move(unpack(args))
+  M.last_move = { func = move, args = args }
 end
 M.goto_previous_start = function(query_strings)
-  move(query_strings, not "forward", "start")
+  local args = { query_strings, not "forward", "start" }
+  move(unpack(args))
+  M.last_move = { func = move, args = args }
 end
 M.goto_previous_end = function(query_strings)
-  move(query_strings, not "forward", not "start")
+  local args = { query_strings, not "forward", not "start" }
+  move(unpack(args))
+  M.last_move = { func = move, args = args }
+end
+
+-- implements naive f, F, t, T with repeat support
+local function builtin_find(forward, inclusive, char, repeating)
+  -- forward = true -> f, t
+  -- inclusive = false -> t, T
+  -- if repeating with till (t or T, inclusive = false) then search from the next character
+  -- returns nil if cancelled or char
+  char = char or vim.fn.nr2char(vim.fn.getchar())
+  repeating = repeating or false
+
+  if char == vim.fn.nr2char(27) then
+    -- escape
+    return nil
+  end
+
+  local line = vim.api.nvim_get_current_line()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+
+  -- find the count-th occurrence of the char in the line
+  local found
+  for _ = 1, vim.v.count1 do
```

Comment:
I know the other plugin used this code too. Could you explain why we can't use `normal! f`?

Reply:
The problem is we want to repeat, without having to enter the same character again.

Reply:
For example if you do `f$` which will find the dollar sign, pressing `;` should  just follow the next dollar sign, not the next function or something else. If we do `normal! f` there's no way to know which character has been searched.

Reply:
Oh sorry, I get your point now. Ignore the comments above. I'll test if it works with normal!

Reply:
The problem is till `t` commands.

From text
```
treesitter
```

You run `tt` and it goes before the second t. Running `tt` again wouldn't change the position, but pressing `;` will move it to the before-third-t.

So, if we want to do `normal! t` then we need to go to the right first with `normal! l` and then performa `normal! t`.
The problem is that if we didn't find anything in that line, it shouldn't move to the right. So we need to detect if it found something. However with `normal! t` it is not easy to detect it and cancel it. Only way is to check the cursor position and then move back which will create an unnecessary cursor move forward and backward.

Reply:
Another issue is, if we want to get the position of the cursor, we need winid. I don't think there's a way to call `vim.cmd[[normal! f]]` with respect to the winid?

Reply:
So the behavior of this is `tt` but without considering the current position. Ok. 

I think you can do run a command also in the context of a certain window. But maybe that's simpler.

Reply:
> So the behavior of this is `tt` but without considering the current position. Ok.
> 
> I think you can do run a command also in the context of a certain window. But maybe that's simpler.

Yes, but only when `repeating=true` to match the vim's behaviour.